tgmusicbot.py

```diff
@@ -96,6 +96,40 @@ def get_music_chats():
 )
 
 
+def extract_link(message: Message):
+    leech_url = None
+    custom_file_name = None
+    entities = (
+        message.entities or
+        message.caption_entities or
+        []
+    )
+    text = (
+        message.text or
+        message.caption or
+        ""
+    )
+    if message and text and len(entities) > 0:
+        for one_entity in entities:
+            if one_entity.type == "url":
+                leech_url = text[
+                    one_entity.offset:one_entity.offset + one_entity.length
+                ]
+            elif one_entity.type == "text_link":
+                leech_url = one_entity.url
+            if leech_url:
+                break
+        if "|" in text:
+            _, custom_file_name = text.split("|", maxsplit=1)
+        if leech_url:
+            leech_url = leech_url.strip()
+        if custom_file_name:
+            custom_file_name = custom_file_name.strip()
+        else:
+            custom_file_name = None
+    return leech_url, custom_file_name
+
```

Comment:
interesting, how did you find that youtube ignores `|.*$` part in the URL?

the whole function seems could just be replaced by this

`yt_url, _, cfn = message.text.partition('|')`

and

1. caption check is useless here because it has `filter.text`. and it's rare case the one send link in a captain instead of just text in text messages.
2. are you sure it should return the full filename? I want to avoid transcoding. and that's why this bot downloads opus/m4a instead of download them and convert to mp3. also that seems it's not possible to include a space character in the user specified file name.
3. checking url entity does not actually check if the URL was valid/available. the only way to check it is sending a HTTP request. user accounts receive preview part which can be used to check webpages, but real bots does not receive that. since it already matches the regex, I don't think it make sense to extract the URL from entities instead of just `message.text`